src/devices/src/virtio/vsock/persist.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines state and support structures for persisting Vsock devices and backends.
+
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use super::*;
+use snapshot::Persist;
+use versionize::{VersionMap, Versionize, VersionizeError, VersionizeResult};
+use versionize_derive::Versionize;
+use vm_memory::GuestMemoryMmap;
+
+use crate::virtio::persist::VirtioDeviceState;
+use crate::virtio::{DeviceState, Queue};
+
+/// The Vsock serializable state.
+#[derive(Versionize)]
+pub struct VsockState {
+    cid: u64,
+    virtio_state: VirtioDeviceState,
+}
+
+/// An enum for the serializable backend state types.
+#[derive(Versionize)]
+pub enum VsockBackendState {
+    Uds(VsockUdsState),
+}
+
+/// The Vsock Unix Backend serializable state.
+#[derive(Versionize)]
+pub struct VsockUdsState {
+    /// The path for the UDS socket.
+    path: String,
+}
+
+/// A helper structure that holds the constructor arguments for VsockUnixBackend
+pub struct VsockConstructorArgs<B> {
+    mem: GuestMemoryMmap,
+    backend: B,
+}
```

Comment:
nit: all devices will also need to take a reference to `EventManager` to register their handlers on `restore()`.

But we can add that incrementally when we get there.

Reply:
I agree that there will be some extra work to kick the restored devices, but I am not sure we will ever need an EventManager ref in the constructor. My best bet is reusing the activate path to register the events.